lsp-pyls.el

```diff
@@ -353,53 +359,84 @@ parameters referenced in config."
   :group 'lsp-pyls
   :package-version '(lsp-mode . "6.3"))
 
+(defcustom lsp-pyls-plugins-jedi-use-pyenv-environment t
```

Comment:
`pyenv` only manages python versions. For project management, you probably use things like `poetry` or `pipenv`  and combined them with `pyenv virtualenv`(that's my workflow though). So having `pyenv` installed doesn't mean the user would definitely like to use it in pyls. I think it's necessary to keep this flag to leave the control for users.

Reply:
What I mean, is that you either manage all of your python versions with `pyenv` (through `.python_version` for project of through `pyenv global`) or you don't use `pyenv` at all. This is different from `virtualenvs`.

In my opinion, virtualenvs should not be managed  in lsp-mode. If you use virtualenvs, there are great emacs packages for them, that set `exec-path` properly. So, imo, the flow should be like this:
1. use `lsp-pyls-plugins-jedi-environment` if it is set.
2. Check if we are in virtualenv by `(getenv "VIRTUAL_ENV")` if yes -> send nil
3. Check if pyenv is present. If yes -> send it
4. In other situations send nil

Reply:
Also, currently you don't check that pyenv is actually installed.

If `lsp-pyls-plugins-jedi-use-pyenv-environment` is `t`, and it is by default, user without pyenv installed will get initialization errors here: https://github.com/emacs-lsp/lsp-mode/pull/1393/files#diff-ce06d48fb3afc95bf7eafe4b4a9f2b11R434

Reply:
I agree with that if `pyenv` is installed then we should at least fallback to `pyenv global`.

I didn't mentioned `virtualenvs` but https://github.com/pyenv/pyenv-virtualenv. I am not familiar how `VIRTUAL_ENV` is set by other emacs packages so I stick with getting the current environment by `pyenv`.

Then for the `exec-path`, it won't help in a use case, which is the goal of this PR: `pyls` is installed globally instead of per python version. Different python environments will be passed into it when executing this global `pyls`. So `exec-path` won't help as the `pyls` is always like `/usr/bin/pyls` and that's why `pyls.plugins.jedi.environment` comes.

I'd like to propose a flow like this(only change the 2nd point):
1. use `lsp-pyls-plugins-jedi-environment` if it is set.
~~2. Check if we are in virtualenv by (getenv "VIRTUAL_ENV") if yes -> **send it**~~(Out of the scope of this PR. Leave if to people who knows how `VIRTUAL_ENV` is set per project in Emacs)
3. Check if pyenv is present. If yes -> send it
4. In other situations send nil

Added the check to see if `pyenv` is installed. Thanks for pointing out.(I assumed that when `lsp-pyls-plugins-jedi-use-pyenv-environment` is `t` user should have `pyenv` installed but forgot it's set to `t` by default.)

Reply:
Will pyenv use its virtualenv by default?

If you don't set `lsp-pyls-plugins-jedi-environment` and you have `pyenv-virtualenv` with pyls installed inside, will this work?
If no, this is a problem, as it will break existing user configurations.
"VIRTUAL_ENV" is set by `virtualenv`, not by specific emacs packages. If `pyenv-virtualenv` wraps around it, it sets it as well

Reply:
1. `pyenv` won't use `virtualenv`.
2. `pyenv-virtualenv` is managed by `pyenv`, like `pyenv global`, `.python_version`, etc.
3. I meant the `VIRTUAL_ENV` of the shell env inside Emacs.

Let's say we have a work flow like this:
1. We have only one pyls binary and `lsp-pyls-server-command` is set to it, i.e. `/usr/bin/pyls`
2. We have two project `proj1` and `proj2` and have `.python_version` specifying `2.7` and `3.7` respectively.(If using `pyenv-virtualenv`, the only difference is `.python_version` will be `env1` and `env2`)
3. We open to these projects in Emacs and call `lsp` on after another

What we expect is that the `.python_version` is regarded for each project automatically and passed to pyls as `pyls.plugins.jedi.environment`. Shell ENV won't help with this. If we want to pass the ENV as `pyls.plugins.jedi.environment`, the ENV need to be set automatically by an Emacs package when we switch buffer. Such packages that I've played with do things manually.

In summary, considering that users have their own ways to config `pyls`, such as:
Write some script to dynamically config `lsp-pyls-server-command`, according to any information from such as the buffer, `(shell-command-to-string "pyenv which)`", `PYENV_VERSION` or `VIRTUAL_ENV` that's populated by https://github.com/pythonic-emacs/pyenv-mode or whatever. It's hard to predict the intended behavior.

Hence I'd like to protect this PR by the `lsp-pyls-plugins-jedi-environment` flag and set it to `nil` by default. Athough I think it's a good way to detect the environment if `pyenv` is installed, let's leave it for users to enable. Python environment management is such a mess. I don't want to think of all the cases that the users want to hack on it.

Reply:
The flow with latest commit will be:
```python
if lsp-pyls-plugins-jedi-environment is set:
  send `pyls.plugins.jedi.environment`
else if lsp-pyls-plugins-jedi-environment is enabled:
  send `pyls.plugins.jedi.environment`
else:
  do nothing and the behavior won't change
```